corehq/apps/data_cleaning/tasks.py

```diff
@@ -42,17 +44,17 @@ def commit_data_cleaning(self, bulk_edit_session_id):
     session.purge_records()
 
     form_ids = []
-    case_index = 0
     session.update_result(0)
     count_cases = case_load_counter("bulk_case_cleaning", session.domain)
-    while case_index < session.records.count():
-        records = session.records.all()[case_index:case_index + CASEBLOCK_CHUNKSIZE]
-        case_index += CASEBLOCK_CHUNKSIZE
-        blocks = _create_case_blocks(session, records)
+
+    record_iter = session.records.order_by('pk').iterator()
+    for record_batch in chunked(record_iter, CASEBLOCK_CHUNKSIZE):
+        blocks = _create_case_blocks(session, record_batch)
         xform = _submit_case_blocks(session, blocks)
```

Comment:
I think I might be confused about the code here. We're iterating over all the records in a session, which likely represent multiple forms, right? And each `record_batch` also likely represents multiple forms? How can we be sure that `_submit_case_blocks(session, blocks)` only creates a single `xform` then?

Reply:
This is a question for @orangejenny, but I believe the answer is that this is a bulk case update form and is already part of the case importer infrastructure